[llvm-profgen] Add support for Linux kenrel profile

xur-llvm opened the pull request:
Add the support to handle Linux kernel perf files. The functionality is under option -kernel. Note that currently only main kernel (in vmlinux) is handled: kernel modules are not handled.

xur-llvm committed:
[llvm-profgen] Add support for Linux kenrel profile

Add the support to handle Linux kernel perf files. The functionality is under
option -kernel. Note that currently only main kernel (in vmlinux) is
handled: kernel modules are not handled.

xur-llvm commented:
Ping.

WenleiHe commented on llvm/tools/llvm-profgen/ProfiledBinary.h:
```diff
@@ -428,6 +431,14 @@ class ProfiledBinary {
   bool useFSDiscriminator() const { return UseFSDiscriminator; }
+  bool isKernel() const { return IsKernel; }
+
+  bool isKernelImageName(StringRef BinaryName) const {
```
Why do we need a `BinaryName` parameter given `ProfiledBinary` has a name associated with it (`getName`)?

xur-llvm replied:
getName() would return the binary name of the kernel image, typically vmlinux, or a name with version info, like vmlinux-6.8.0-smp-1201.1.889.0. The image name from perf command is in of the forms in this function.

WenleiHe replied:
Ok, then this is static function?

xur-llvm replied:
Right. I'll make it static.

WenleiHe commented on llvm/tools/llvm-profgen/ProfiledBinary.cpp:
```diff
@@ -56,6 +56,10 @@ static cl::list<std::string> DisassembleFunctions(
              "names only. Only work with show-disassembly-only"));
 
+static cl::opt<bool>
+    KernelBinary("kernel",
```
Given we assume kernel binary has specific names, do we need an extra flag to tell the tool to generate profile for kernel? 

If we have this flag, what happens if the flag is used but binary provided doesn't match the predefined kernel binary name? Should we error out?

xur-llvm replied:
I think this interface is easier than parsing the binary name. Usually the kernel image name is vmlinux.*. But there is a chance that an user image uses the same name. Or someone changes the kernel image name to whatever. I want users to confirm this is kernel perf data file by using the option.

If the user passes a wrong binary image, most like they won't get a profile because of the symbol and address mismatch.

Compared to AutoFDO's create_llvm_prof, llvm-profgen does not check the buildid. But that's separated issue. This patch is just the basic version that provides the kernel perf support.

WenleiHe replied:
Ok, then this should be a static function?

WenleiHe replied:
> If the user passes a wrong binary image, most like they won't get a profile because of the symbol and address mismatch.

Maybe useful to error out if kernel is requested but binary is not kernel?

xur-llvm replied:
That actually what I want to avoid -- I don't want to assume a fixed name for the kernel. For my kernel builds, I don't change anything and the kernel image is vmlinux.*. But I don't know if others will rename the image.

WenleiHe replied:
What I meant is error out when kernel image isn't found in mmap (we're already assuming names there). But no big deal either way.

shenhanc78 replied:
Done.

wlei-llvm commented on llvm/tools/llvm-profgen/PerfReader.cpp:
```diff
@@ -355,57 +355,69 @@ PerfReaderBase::create(ProfiledBinary *Binary, PerfInputFile &PerfInput,
-      if (It.second && (!PIDFilter || MMap.PID == *PIDFilter)) {
-        if (!PIDs.empty()) {
-          PIDs.append(",");
+  if (!Binary->isKernel()) {
```
How about making this a function parameter and pass it from the caller? thinking this may be useful for other binary to skip the PID passing.

xur-llvm replied:
Sure. I'll do that.

shenhanc78 replied:
Thanks. @wlei-llvm Done (by passing a parameter instead of call Binary->IsKernel()) here.

xur-llvm committed:
[llvm-profgen] Add support for Linux kenrel profile

Integrated Wenlei's review comments:
(1) make retrieving the PID a function parameter.
(2) make function isKernelImageName() a static function.

shenhanc78 committed:
[llvm-profgen] Add support for Linux kenrel profile

Integrated Wenlei's review comments:
(1) make retrieving the PID a function parameter.
(2) make function isKernelImageName() a static function.
(3) error out if "--kernel" is given, but no corresponding mmap event is found
(4) change "Binary->isKernel()" to a function parameter

shenhanc78 committed:
[llvm-profgen] Add support for Linux kenrel profile

Integrated Wenlei's review comments:
(1) make retrieving the PID a function parameter.
(2) make function isKernelImageName() a static function.
(3) error out if "--kernel" is given, but no corresponding mmap event is found
(4) change "Binary->isKernel()" to a function parameter

shenhanc78 commented:
Since @xur-llvm is on vacation till Jun, 25. I'll take over here. Thanks.

shenhanc78 committed:
[llvm-profgen] Add support for Linux kenrel profile

Integrated Wenlei's review comments:
(1) make retrieving the PID a function parameter.
(2) make function isKernelImageName() a static function.
(3) error out if "--kernel" is given, but no corresponding mmap event is found
(4) change "Binary->isKernel()" to a function parameter

WenleiHe reviewed (approved):
lgtm, thanks.
